types/wlr_surface.c

```diff
@@ -365,12 +378,14 @@ static void surface_apply_damage(struct wlr_surface *surface) {
 		}
 	}
 
-	struct wlr_client_buffer *buffer =
-		wlr_client_buffer_import(surface->renderer, resource);
+	struct wlr_client_buffer *buffer = wlr_client_buffer_create(
+			surface->current.buffer, surface->renderer, resource);
 	if (buffer == NULL) {
 		wlr_log(WLR_ERROR, "Failed to upload buffer");
 		return;
 	}
+	wlr_buffer_unlock(surface->current.buffer);
+	surface->current.buffer = NULL;
```

Comment:
Here we take care of unlocking the buffer. But we don't do it in all branches of this function. We should also unlock the buffer above when `return` is used.